README.md
# Under the Xadian Sky

A simple game based on the Netflix series "The Dragon Prince"

## Changes
### Latest version: 0.1.3 - 2020-05-25
#### Added
- Simpler save file validation
- Function to check if save file is valid
- Progress log in [debug.py](debug.py)
- Name input for invalid save files
- [LICENSE](LICENSE) (GNU General Public License v3)
- [README.md](README.md)
- [Github](https://github.com/the2nake/under-the-xadian-sky) remote repository and [Git](https://git-scm.com) source control


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
 - Made module "source", to store code for parts of the program

## [0.2.0]
### Added
 - New icon for the game window
 - Logging files for debugging
 - Fixed save file checks
 - Can now save the avatar, implemented simple file-saving function
 - Reformatted this file (changed version numbering)

## [0.1.2] - 2020-05-25
## Added
 - Simpler save file validation
 - Function to check if save file is valid
 - Progress log in [debug.py](debug.py)
 - Name input for invalid save files
 - [LICENSE](LICENSE) (GNU General Public License v3)
 - [README.md](README.md)
 - [Github](https://github.com/the2nake/under-the-xadian-sky) remote repository and [Git](https://git-scm.com) source control

## [0.1.1] - 2020-05-10
### Added
 - Changes in [tilemap.py](tilemap.py) to add more tiles in the tilesheet
 - A revamped introduction screen, to select saves first
 - Beginning save file implementation and validation
 - Various functions and classes in pydraw.py to assist in drawing with pygame
 - Updated dependencies to pygame 2.0.0.dev8 because of border-radius
 - Added multiple docstrings and comments

## [0.1.0] - 2020-04-28
### Added
 - [tilemap.py](tilemap.py) - a module defining names and positions for the tileset [tilesheet.png](assets/tilesheet.png)
 - A semi-working introduction screen

pydraw.py
''' 
Pydraw by Vo Thuong
-------------------
A collection of modules that help with drawing in pygame

Dependecies:
pygame 2.0.0.dev8
tilemap.py
'''

import pygame
import tilemap
from pygame import freetype


class Button:
    '''
    A simple button class for pygame

    Arguments:
    text (str): The text to display on the button

    pos, size, colour (tuple): The position, size, and colours of the button, drawn from the top left corner of the button. The pos and size support using sw, sh, bw, bh to stand for the screen and button's height and width. Default colours are (255, 0, 0) when not clicked and (128, 0, 0) when clicked.

    icon (tuple): Formatted like so: (tilesheetname (str), iconname (str, in tilesheet), pos (bool)). pos is left(f) or right(t). If iconname = None or invalid, no icon.
    disabled (bool): If the button is "fake" or not
    borderradius (int): The radius of the border. Default is 0

    Note: the colour tuple is a tuple of 3 tuples, both representing colours. The first is non-clicked, the second is clicked, the  third is text colour.
    '''

    def __init__(self, screen: pygame.Surface, text: str, pos: tuple, size: tuple, colours: tuple = ((255, 0, 0), (128, 0, 0), (0, 0, 0)), font: tuple = (None, 24), icon: tuple = ("assets/tilesheet.png", None, False), disabled=False, borderradius: int = 0):
        sw = screen.get_width()
        sh = screen.get_height()
        bw = size[0]
        bh = size[1]
        if type(pos[0]) == str or type(size[0]) == str:
            self.x = eval(pos[0])
            self.y = eval(pos[1])
        else:
            self.x = pos[0]
            self.y = pos[1]

        if type(size[0]) == str:
            self.w = eval(size[0])
            self.h = eval(size[1])
        else:
            self.w = size[0]
            self.h = size[1]
            
        self.rect = pygame.rect.Rect(
            int(self.x), int(self.y), int(self.w), int(self.h))
        self.screen = screen
        self.font = font
        self.text = text
        self.icon = icon
        self.colours = colours
        if disabled:
            self.drawcolour = colours[1]
        else:
            self.drawcolour = colours[0]
        self.borderradius = borderradius
        self.disabled = disabled

    def handle_event(self, event):
        '''
        Handles the clicking
        '''
        if self.disabled:
            self.drawcolour = self.colours[1]
            return 0
        else:
            if event.type == pygame.MOUSEBUTTONDOWN:
                # If the user click on the button
                if self.rect.collidepoint(event.pos):
                    self.drawcolour = self.colours[1]
                    return 1
            elif event.type == pygame.MOUSEBUTTONUP:
                self.drawcolour = self.colours[0]
                return 0

            return None

    def draw(self):
        '''
        Draws the button
        '''
        #pygame.draw.rect(self.screen, self.drawcolour, self.rect, 0, self.borderradius)
        pygame.draw.rect(self.screen, self.drawcolour, self.rect, 0, self.borderradius)
        if self.icon[1] != None and tilemap.names.__contains__(self.icon[1]):
            padding = self.h/2 - 8
            if self.icon[2]: # right
                draw_tile(self.screen, str(self.icon[0]), (self.x + self.w - 16 - padding, self.y + padding), tilemap.init()[0][self.icon[1]], 0, (False, False))
                text(self.screen, self.text, ("%s + (%s - tw - 16 - 2 * %s)/2" % (str(self.x), str(self.w), str(padding)),
                                              "%s + (%s - th)/2" % (str(self.y), str(self.h))), self.colours[2], self.drawcolour, self.font)
            else: # left
                draw_tile(self.screen, str(self.icon[0]), (self.x + padding, self.y + padding), tilemap.init()[0][self.icon[1]], 0, (False, False))
                text(self.screen, self.text, ("%s + 24 + (%s - tw - 16 - 2 * %s)/2" % (str(self.x), str(self.w), str(padding)),
                                                "%s + (%s - th)/2" % (str(self.y), str(self.h))), self.colours[2], self.drawcolour, self.font)
        else:
            text(self.screen, self.text, ("%s + (%s - tw)/2" % (str(self.x), str(self.w)),
                                          "%s + (%s - th)/2" % (str(self.y), str(self.h))), self.colours[2], self.drawcolour, self.font)


class InputBox:
    '''
    A simple input box class for pygame

    Arguments:

    x, y, w, h (int): pos, size of input box

    colours (tuple): tuple of tuple of colours. ((active), (inactive), (bgcolour))
    '''

    def __init__(self, x: int, y: int, w: int, h: int = 24, colours: tuple = ((168, 70, 70), (150, 50, 50), (0, 0, 0)), deftext=''):
        self.rect = pygame.Rect(x, y, w, h)
        self.keymap = {}
        self.colour = pygame.Color(colours[1][0], colours[1][1], colours[1][2])
        self.colours = colours
        self.text = deftext
        self.deftext = deftext
        self.font = h
        self.txt_surface = pygame.font.Font(
            None, self.font).render(self.text, True, self.colour)
        self.active = False

    def handle_event(self, event):
        '''
        Handles the events related to this input box
        '''
        if event.type == pygame.MOUSEBUTTONDOWN:
            # If the user clicked on the input_box rect.
            if self.rect.collidepoint(event.pos):
                # Toggle the active variable.
                self.active = not self.active
            else:
                self.active = False
            # Change the current color of the input box.
            self.colour = pygame.Color(self.colours[0][0], self.colours[0][1], self.colours[0][2]) if self.active else pygame.Color(
                self.colours[1][0], self.colours[0][1], self.colours[1][2])
        if event.type == pygame.KEYDOWN:
            self.keymap[event.key] = True
            if self.active:
                if self.keymap.get(pygame.K_RETURN):
                    if self.text == "":
                        self.deftext = "Input may not be empty"
                        self.text = self.deftext
                        self.active = False
                    else:
                        return self.text
                elif self.keymap.get(pygame.K_BACKSPACE):
                    self.text = self.text[:-1]
                else:
                    self.text += event.unicode
                # Re-render the text.
                self.txt_surface = pygame.font.Font(
                    None, self.font).render(self.text, True, self.colour)
        if event.type == pygame.KEYUP:
            self.keymap[event.key] = False

        return 0

    def update(self):
        '''
        Updates the input box
        '''
        # Resize the box if the text is too long.
        width = max(self.rect.w, self.txt_surface.get_width()+10)
        self.rect.w = width

        if self.active and self.text == self.deftext:
            self.text = ''
            self.txt_surface = pygame.font.Font(
                None, self.font).render(self.text, True, self.colour)
        elif (not self.active) and self.text == '':
            self.text = self.deftext
            self.txt_surface = pygame.font.Font(
                None, self.font).render(self.text, True, self.colour)

        return 0

    def draw(self, screen: pygame.Surface):
        '''
        Draws the input box
        '''
        # Blit the text.
        screen.blit(self.txt_surface, (self.rect.x+5, self.rect.y+5))
        # Blit the rect.
        pygame.draw.rect(screen, self.colour, self.rect, 2)

        return 0


def text(screen: pygame.Surface, text: str, pos: tuple, colour: tuple, bgcolour: tuple = (0, 0, 0), font: tuple = (None, 24)):
    '''
    Blit text to the screen, with varying levels of transparency.
    ---
    Note: This function supports the use of expressions to evaluate the position for the text. The variables are:
    textw, texth, scrnw, scrnh for text width and height and screen width and height

    Arguments:
    ---
    screen (pygame.Surface): The screen surface. \n
    text (string): text to render \n
    pos (tuple): A tuple describing where on the screen should the text be drawn. \n
    colour, bgcolor (tuple): Format: (r, g, b, (a)), where 0 <= r, g, b, a <= 255. \n
    font (tuple): Format: (path to the font file, fontsize). The pygame freetype module supports TTF, Type1, CFF, OpenType, SFNT, PCF, FNT, BDF, PFR and Type42 fonts. Default is (None, 32)
    '''
    if len(colour) == 4:
        mode = "rgba"
    else:
        mode = "rgb"
    rgbcolour = pygame.Color(colour[0], colour[1], colour[2])

    if font[0] == None:
        textobj = pygame.font.Font(None, font[1])
        textobj = textobj.render(text, True, rgbcolour)
        tw = textobj.get_width()
        th = textobj.get_height()
    else:
        textobj = freetype.Font(font[0], font[1])
        textobj = textobj.render(text, rgbcolour)
        tw = textobj[0].get_width()
        th = textobj[0].get_height()


    sw = screen.get_width()
    sh = screen.get_height()
    rendering_surface = pygame.Surface((tw, th))
    try:
        rendering_surface.fill(pygame.Color(
            bgcolour[0], bgcolour[1], bgcolour[2]))
    except IndexError:
        return None
    
    if font[0] == None:
        rendering_surface.blit(textobj, (0, 0))
    else:
        rendering_surface.blit(textobj[0], (0, 0))

    # pygame.font.Font().render's alpha cannot be set, use surface alpha
    if mode == "rgba":
        if 0 <= colour[3] <= 255:
            rendering_surface.set_alpha(colour[3])
        elif colour[3] > 255:
            rendering_surface.set_alpha(255)
        else:
            rendering_surface.set_alpha(0)

    if isinstance(pos[0], str) and isinstance(pos[1], str):

        # evaluate expressions
        screen.blit(rendering_surface, (int(eval(pos[0])), int(eval(pos[1]))))

    else:
        try:
            screen.blit(rendering_surface, (pos[0], pos[1]))
        except IndexError:
            return None

    return 1


def draw_img(surface: pygame.Surface, img_path: str, _x: int, _y: int):
    """
    Draws an image with source imgpath at (x, y), on the surface surface

    Arguments:
    surface (pygame.Surface): the pygame surface to draw on
    imgpath (string): the path to the image
    x (int): where to the image (x-axis)
    y (int): where to the image (y-axis)
    """
    try:
        img = pygame.image.load(img_path)
    except pygame.error:
        return 0

    try:
        surface.blit(img, (_x, _y))
    except TypeError:
        return 0

    return 1


def draw_tile(surface: pygame.Surface, img_path: str, pos: tuple, area: tuple, angle: int = 0, flip: tuple = (False, False)):
    """
    Draws an image with source imgpath at (x, y), on the surface surface

    Arguments:
    surface (pygame.Surface): the pygame surface to draw on
    imgpath (string): the path to the image
    pos (tuple): 2-item, position on screen
    area (tuple): 4-item, area to crop from image. (top, left, width, height) format
    angle (int): 0, 1, 2, or 3. Rotates in 90-degree increments
    flip (tuple): (horz, vert) flips. Applied after rotation
    """
    try:
        img = pygame.image.load(img_path)
    except pygame.error:
        return 0

    cropped = pygame.transform.chop(img, pygame.rect.Rect(
        0, 0, area[0], area[1]))  # top, left
    cropped = pygame.transform.chop(cropped, pygame.rect.Rect(
        area[2], area[3], 543 - area[0], 543 - area[1]))  # bottom, right
    rotated_image = pygame.transform.rotate(
        cropped, 90*(4-angle))  # reversed direction
    final_image = pygame.transform.flip(rotated_image, flip[0], flip[1])
    # new_rect = rotated_image.get_rect(
    #    center=cropped.get_rect(topleft=(0, 0)).center)

    #surface.blit(rotated_image, (new_rect.topleft[0] + pos[0], new_rect.topleft[1] + pos[1]), area)

    surface.blit(final_image, (pos[0], pos[1]))

    return 1


tilemap.py
'''
Module defining tileset for the game "Under a Xadian Sky"

Dependecies: None
'''

names = "empty dirt1 dirt2 dirt3 dirt4 grass1 grass2 grass3 road road-corner road-three-intersection road-intersection road-end thin-wall thin-wall-with-bump thin-wall-corner brick-tile kitchen-tile top-left-stone-floor top-middle-stone-floor top-right-stone-floor ladder-end spike alert-outline person-raincoat person1 person-spearman person-swordsman person-crusader person-pikeman person-armoured person-warlord "


names = names + "tree1 tree2 tree3 tree4 tree5 tree6 cactus cacti path path-corner path-three-intersection path-intersection path-end lane lane-corner lane-three-intersection lane-intersection footsteps center-left-stone-floor platform-stone center-right-stone-floor ladder shaft alert-fill person-great-mage person-boy-scout person-princess person-office person-grunt person-teen person-apprentice person-mage "


names = names + "tree7 tree8 tree9 tree10 great-tree stones dead-tree tree11 carpet carpet-corner carpet-three-intersection carpet-intersection carpet-end pillar-end pillar-wall pillar-corner pillar-intersection pillar-three-intersection bottom-left-stone-floor bottom-middle-stone-floor bottom-right-stone-floor dry-grass shaft-barred box person-elder troll-whelp troll devil pirate yoda person-fencer person-farmer "


names = names + "fence-low fence fence-flat door-gate-closed door-gate-open bars bars-broken indoor-wall indoor-outer-corner indoor-inner-corner indoor-wall-small-hallway indoor-wall-dead-end indoor-wall-with-inner-corner wall-end wall wall-corner wall-intersection wall-three-intersection top-left-stone top-right-stone slant-top-left-stone-floor slant-top-right-stone-floor box-chiseled chessboard person-rogue robot-single-wheel robot robot-strong king queen boy girl "


names = names + "fence-metal-low fence-metal fence-metal-broken door-gate-metal-closed door-gate-metal-open gate-metal-closed bridge-metal-elevated bridge-metal river river-turn river-three-intersection river-intersection river-end wall-outline-end wall-outline wall-outline-corner wall-outline-intersection wall-outline-three-intersection bottom-left-stone bottom-right-stone slant-bottom-left-stone-floor slant-bottom-right-stone-floor coin diamond person-hooded person-old engineer person-wrestler person-businessman person-wheelchair person2 person3 "


names = names + "rail rail-corner rail-three-intersection rail-intersection rail-broken bar-ground bridge-wooden-elevated bridge-wooden river-empty river-shore river-outer-corner river-inner-corner stream stream-turn stream-reservoir roller-coaster-ramp roller-coaster-track roller-coaster-track-broken platform leaves1 leaves2 spring1 spring2 spring3 lobster crab bee turtle spider1 spider2 spider3 wasp "


names = names + "rail-wooden rail-wooden-broken stairs-up stairs-down drawbridge bookshelf arch-wooden unknown1 chest-open chest-closed basket-open basket-closed crossing-side plants1 plants2 plants3 plants4 plants5 tree-trunk1 tree-trunk2 tree-trunk3 platform-start platform-middle platform-end person-dark-mage person-dark ghost1 ghost2 ghost3 skeleton giant1 dark-warrior "


names = names + "sign-direction sign mailbox bookcase bookcase-small wardrobe-closed wardrobbe-open cupboard-closed cupboard-open cupboard-hanging-closed cupboard-hanging-open sink refridgerator refridgerator-open sign-bed sign-semi-circle sign-rectangle banner blue-run1 blue-run2 blue-run3 blue-jump1 blue-jump2 blue-die person-empty-with-face duck chicken cow horse pig cat dog "


names = names + "mirror-standing bedtable table-up table-side-end table-side-middle bed bed-start bed-end chimney fireplace pancakes stove1 stove2 basin sign-hanging-bed sign-handing-semicircle table-up-end flag yellow-run1 yellow-run2 yellow-run3 yellow-jump1 yellow-jump2 yellow-die person-empty octopus bat blob snake crocodile bear rat "


names = names + "door-lock door-energy-field door-frame-blue door1 door2 door3 door-frame door4 door5 door-frame-thin door6 chair-side1 chair-up1 toilet-up table-side1 table-side2 table-up-middle flag-triangle green-run1 green-run2 green-run3 green-jump1 green-jump2 green-die cyclops giant2 gnome person-assasin person-theif person-warior person-general mermaid "


names = names + "lock button button-pressed lever-off lever-on door-metal1 door-metal2 door-metal-garage doorframe-metal door-metal-end door-metal-garage-end doorframe-metal-end toilet-side bathtub-side campfire fire table-up-start1 table-up-start2 house-green tent-green sign-asian lantern-hanging lights-hanging bell avatar-boy avatar-girl avatar-man-old avatar-man-bald-man avatar-man avatar-man2 avatar-girl2 avatar-woman "


names = names + "lock-square stairs-stone-up wall-stone-top castle-patrol-up-top bar-locked roof-tile1 roof-tile2 roof-tile3 door-metal-bars1 door-metal3 door-metal4 door-metal5 door-metal-bars-open door-metal-bars2 sign-road tires-stack drawbridge-side-open door-blocked roof-green-start roof-green-middle roof-green-end asian-arch1 asian-arch2 gong gfx-slash gfx-slash-arch gfx-scratch gfx-magical gfx-fire gfx-missile gfx-slash-x gfx-burn "


names = names + "castle-tower-top1 castle-tower-top2 castle-tower-emblem castle-patrol-up-middle table-locked roof-start roof-middle roof-end pipe-end pipe-middle pipe-invalid pipe-support chain idol wires-hanging tire rubble gfx-bubbles curtain-end window-bars1 door7 window-goemetric window-bars2 crossing-up arrow-up1 sectioning-intersection sectioning-intersection-arrows gfx-tree gfx-explode1 gfx-explode2 gfx-confuse gfx-explode3 " 


names = names + "stone-wall stone-window stone-window-small castle-patrol-up-end stone-arch brick-window1 brick-wall1 brick-window2 road-support1 road-support2 crane-beam crane-support hook hard-disks flask crane-wheels castle-wall-top-left castle-wall-top-middle castle-wall-top-right castle-patrol-side-end castle-patrol-side-middle hallway-side-end1 hallway-side-middle hallway-side-end2 arrow-up2 sectioning sectioning-end-curved sectioning-end-fade sectioning-corner-curved sectioning-three-intersection-curved sectioning-corner sectioning-three-intersection"


names = names.split()


def init():
    '''
    This function will initiaze the tilemap

    Arguments: None

    Returns:
    tilemap (dict): The tilemap for every tile on the tilesheet
    '''

    # row 1

    tilemap = {}  # format: (posx, posy, width, height)
    avatarnames = []

    for i, item in enumerate(names):
        tilemap[item] = (17 * (i % 32), 17 * (i // 32), 16, 16)

    
    for name in names:
        if name[0:6] == "avatar":
            avatarnames.append(name)

    return tilemap, (names, avatarnames)


if __name__ == "__main__":
    tilesheet, tilenames = init()
    print(tilesheet.get("empty"))
    print(list(tilesheet.keys())[64])


testpygame.py
import pygame

pygame.init()
screen = pygame.display.set_mode((640, 360))

ticks = 0
clock = pygame.time.Clock()
introrunning = True
backgroundcolour = (128, 255, 128)
while introrunning:
    for event in pygame.event.get():
        # if the event is quit --> stop
        if event.type == pygame.QUIT:
            introrunning = False

    ticks += 1

    screen.fill(backgroundcolour)
    # pygame.font.Font().render's alpha cannot be set
    welcomescreen = pygame.font.Font(None, 56, flags=pygame.SRCALPHA).render(
        "Under the Xadian Sky", True, pygame.Color(231, 169, 132))
    textsurface = pygame.Surface(
        (welcomescreen.get_width(), welcomescreen.get_height()))
    textsurface.fill(backgroundcolour)
    textsurface.blit(welcomescreen, (0, 0))
    if ticks < 255:
        textsurface.set_alpha(ticks)
    screen.blit(textsurface, (int((screen.get_width() - welcomescreen.get_width()
                                   ) / 2), int((screen.get_height() - welcomescreen.get_height()) / 3)))
    pygame.display.update()
    clock.tick(30)


release.py
"""
This is the entry point for the game "Under the Xadian Sky", based on the
Netflix series "The Dragon Prince".

Dependencies:
os
pygame 2.0.0.dev8
logzero
datetime
pydraw.py
tilemap.py

Thank you kenney, for making this possible.
"""

import os
import pygame
import tilemap
import logzero
import datetime
from pydraw import *
from logzero import logger

# Input box class

now = datetime.datetime.now()
date = now.strftime("%d-%m-%Y")
time = now.strftime("%H:%M:%S")


def checksave(savefile: str):
    savedata = {}
    try:
        file = open(savefile)
    except FileNotFoundError:
        return False, {}
    savelist = file.readlines()
    tilesheet, tilenames = tilemap.init()
    
    try:
        for each in savelist:
            each = each[:len(each) - 1]
    
        if savelist[0].split("=")[0] == "name" and len(savelist[0].split("=")[1]) > 2 and savelist != [] and savelist[1].split("=")[0]== "avatar" and len(savelist[1].split("=")[1]) > 2 and tilenames.__contains__(savelist[1].split("=")[1][:-1]):
            savedata["valid"] = True
            savedata["name"] = savelist[0].split("=")[1][:-1]
            savedata["avatar"] = savelist[1].split("=")[1][:-1]
        else:
            savedata["valid"] = False
            # clear all empty save files
            
    except IndexError:
        savedata["valid"] = False
        
    if savedata["valid"]:
        return True, savedata
    else:
        return False, {}


def main():
    '''
    The main function.
    '''

    logzero.logfile("logs/logfile-" + date)

    logger.info("Game started on %s at %s" % (date, time))
    logger.info("--------------------------------------")

    # init
    pygame.init()

    # create tilesheet for the assets/tilesheet.png file
    tilesheet, tilenames = tilemap.init()

    # create window and setup
    screen = pygame.display.set_mode((800, 600))
    pygame.display.set_caption("Under the Xadian Sky")
    icon = pygame.image.load("icon.png")
    pygame.display.set_icon(icon)
    #backgroundcolour = (128, 255, 128)
    backgroundcolour = (0, 0, 0)
    # speed optimizations
    pygame.event.set_allowed([pygame.QUIT, pygame.KEYDOWN, pygame.KEYUP, pygame.MOUSEBUTTONDOWN, pygame.MOUSEBUTTONUP])
    # pygame.mouse.set_cursor(*pygame.cursors.tri_left)

    pygame.mouse.set_visible(False)
    cursorimg = pygame.image.load("cursor.png")

    # create keymap
    keymap = {}

    clock = pygame.time.Clock()
    input_boxes = []
    buttons = []

    # save-file

    savelists = [[], [], []]
    savedata = [{}, {}, {}]
    savechosen = -1
    savevalidnums = []


    if checksave("save1.txt")[0]:
        savevalidnums.append(1)
        savedata[0] = checksave("save1.txt")[1]
    else:
        savedata[0] = {"valid": False}
        
    if checksave("save2.txt")[0]:
        savevalidnums.append(2)
        savedata[1] = checksave("save2.txt")[2]
    else:
        savedata[1] = {"valid": False}
        
    if checksave("save3.txt")[0]:
        savevalidnums.append(3)
        savedata[2] = checksave("save3.txt")[2]
    else:
        savedata[2] = {"valid": False}

    # save: check validity
    logger.debug("Save file info: %s" % savedata)

    # if not (savedata[0]["valid"] or savedata[1]["valid"] or savedata[2]["valid"]):
    #    input_boxes.append(
    #        InputBox(100, 288, 600, 24, ((168, 61, 61), (150, 50, 50), backgroundcolour), "What is your name? (type here)"))

    # intro init

    introrunning = True
    # there are 2 intro phases: select save, if save is there, open save. If not, ask for name and start a new game
    introphase = "selectsave"
    ticks = 0

    for num in [1, 2, 3]:
        # not savedata[num - 1]["valid"] instead of False for ony loading valid saves
        buttons.append(Button(screen, "Save " + str(num), ("(sw - bw)/2", "(sh - bh)/2 + 48 * %s - 48" % str(num)), (200, 32),
                              ((255, 128, 128), (220, 90, 90), (255, 255, 255)), (None, 24), ("assets/tilesheet.png", savedata[num - 1]["avatar"], False) if savedata[num - 1]["valid"] else ("assets/tilesheet.png", "person-empty", False), False, 4))
        
        # ("assets/tilesheet.png", savedata[num - 1]["avatar"], False) if savedata[num - 1]["valid"] else ("person-empty", False)

    # mainloop
    while introrunning:
        # poll for events
        for event in pygame.event.get():
            # if the event is quit --> stop
            if event.type == pygame.QUIT:
                introrunning = False

            for i, box in enumerate(input_boxes):  # handle inbut boxes
                temp = box.handle_event(event)
                if temp != 0:
                    name = temp
                    logger.info("Name is %s" % name)
                    del input_boxes[i]
                    introrunning = False

            for buttonnum, button in enumerate(buttons):
                out = button.handle_event(event)
                if out == 1:
                    savechosen = buttonnum + 1
                    logger.info("Save chosen: save " + str(savechosen))
                    if introphase == "selectsave":
                        if savedata[savechosen - 1]["valid"]:
                            introphase = "loadsave"
                        else:
                            introphase = "inputname"
                
                    for i, button in enumerate(buttons):
                        if buttons.__contains__(button):
                            buttons.remove(button)
                        else:
                            del buttons[i]

            if event.type == pygame.KEYDOWN:
                keymap[event.key] = True

            if event.type == pygame.KEYUP:
                keymap[event.key] = False

        for box in input_boxes:
            if ticks > 71:
                box.update()
        # set background (grass green)
        screen.fill(backgroundcolour)
        # screen.fill((0, 0, 0))  # rgb

        # Title, credits, author info, etc
        # if savedata[0]["valid"] or savedata[1]["valid"] or savedata[2]["valid"]:  # savefiles not found
        #    text(screen, "Welcome back", ("(sw - tw)/2", "(sh - th)/3 + 50"),
        #         (231, 169, 132, ticks * 5 - 100), backgroundcolour, (None, 32))
        # else:
        #    text(screen, "Save file not found found", ("(sw - tw)/2", "(sh - th)/3 + 50"),
        #         (255, 89, 89, ticks * 5 - 100), backgroundcolour, (None, 32))

        text(screen, "Under the Xadian Sky", ("(sw - tw)/2", "(sh - th)/3"),
             (231, 169, 132, ticks * 5), backgroundcolour, (None, 56))

        if introphase == "selectsave":
            for button in buttons:
                button.draw()
        elif introphase == "inputname":
            if len(input_boxes) == 0:
                input_boxes.append(InputBox(100, 288, 600, 24, ((
                    168, 61, 61), (150, 50, 50), backgroundcolour), "What is your name? (type here)"))
        elif introphase == "loadsave":
            # this is when you load the save. Process tree below:
            #              If save is valid    /----------------------\
            # wait until button is pressed =\/                         \--> load save, start game, introtrunning = False
            #          If save is not valid  \-- Input name, avatar ---/
            pass

        # draw an icon
        # draw_tile(screen, 'assets/tilesheet.png', (304, 164), tilesheet.get("grass1"))

        # Note: slows down window considerably
        #for _ in range(len(tilesheet.keys())):
        #    draw_tile(screen, 'assets/tilesheet.png', (16 * (_ % 32), 16 *
        #                                               (_ // 32)), tilesheet.get(list(tilesheet.keys())[_]), 0, (False, False))
        # draw_tile(screen, 'assets/tilesheet.png', (0, 0),
        #          tilesheet.get(list(tilesheet.keys())[2]), 1)
        # screen.blit(pygame.image.load("assets/tilesheet.png"), (0, 0))

        # check keys
        # if keymap.get(pygame.K_SPACE):
        #    draw_img(screen, 'solicon.png', 304, 164)

        # draw input boxes
        for box in input_boxes:
            if ticks > 71:
                box.draw(screen)

        # draw cursor
        screen.blit(cursorimg, (pygame.mouse.get_pos()))

        if ((savedata[0]["valid"] or savedata[1]["valid"] or savedata[2]["valid"]) and 0 <= savechosen <= 2 and ticks > 100) or (not introrunning):
            introrunning = False
            fade = pygame.Surface((screen.get_width(), screen.get_height()))
            fade.fill((0, 0, 0))
            for alpha in range(0, 40):
                fade.set_alpha(alpha)
                screen.blit(fade, (0, 0))
                pygame.display.update()
                clock.tick(30)
            # pygame.quit()
        else:
            # Update screen
            pygame.display.update()
            ticks += 1
            clock.tick(30)


main()
